### PR TITLE
kvm: load required kernel modules if necessary

### PR DESCRIPTION
**Pull request by gerboland**
Have the KVM interface load the suitable KVM kernel modules.

Supports checking the required CPU capabilities for Intel and AMD CPUs and loading suitable modules if supported, falls back to generic KVM module otherwise.

Ref: https://www.linux-kvm.org/page/Processor_support

### CODE REVIEW

**Comment by gerboland** [comment]
Spread failed for some other reason:

```
Cloning into '/home/gopath/.cache/govendor/golang.org/x/crypto'...
fatal: unable to access 'https://go.googlesource.com/crypto/': Failed to connect to go.googlesource.com port 443: Connection timed out
Error: Remotes failed for:
	Failed for "golang.org/x/crypto/cast5" (failed to clone repo): exit status 128
```

File: interfaces/builtin/kvm.go [hunk 1]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+	if err != nil {
+		// if we can't read cpuinfo, we want to know _why_
+		return "", fmt.Errorf("error: %v", err)
+	}
+
+	pattern := []byte("\nflags\t\t:")
+	idx := bytes.LastIndex(buf, pattern)
+	if idx >= 0 {
+		offset := idx + len(pattern)
+		endidx := bytes.Index(buf[offset:], []byte("\n"))
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: This can return `-1`, it's not super likely but we should check for it.

File: interfaces/builtin/kvm.go [hunk 2]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+		return string(buf[offset : endidx+offset]), nil
+	}
+
+	// if not found (which will happen on non-x86 architectures, which is ok
+	// because they'd typically not have the same info over and over again),
+	// return whole buffer; otherwise, return from just after the \n
+	return string(buf[idx+1:]), nil
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: I don't think this is sensible behaviour. Just return an empty string in this case.

File: interfaces/builtin/kvm.go [hunk 3]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+}
+
+func (iface *kvmInterface) KModConnectedPlug(spec *kmod.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Check CPU capabilities to load suitable module
+	// TODO: this only considering x86 CPUs, but some ARM, PPC and S390 CPUs also support KVM
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: Please at the very least support x86_64, i386 and amd64. I don't know of any easily available ARMv7 boards that have kvm (none of the reference platforms do) so I think it is fine to skip that.

File: interfaces/builtin/kvm.go [hunk 4]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+	m := "kvm"
+	cpu_flags, err := getCpuFlags()
+	if err != nil {
+		fmt.Println("kvm: fetching cpu info failed:", err)
+	}
+
+	if strings.Contains(cpu_flags, "vmx") {
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: I would feel better if cpu_flags was `[]string` or if this was entirely encapsulated in a helper function.

File: interfaces/builtin/kvm.go [hunk 5]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+		m = "kvm_intel"
+	} else if strings.Contains(cpu_flags, "svm") {
+		m = "kvm_amd"
+	} else {
+		// CPU appears not to support KVM extensions, fall back to bare kvm module as it appears
+		// sufficient for some architectures
+		fmt.Println("kvm: failed to detect CPU specific KVM support, will attempt to modprobe generic KVM support")
```
Comment (zyga) [inline interfaces/builtin/kvm.go]: Please use a logger, don't print this way.

File: interfaces/builtin/kvm_test.go [hunk 6]
```diff
@@ -57,8 +66,23 @@ slots:
...
 
 func (s *kvmInterfaceSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+
 	s.plug, s.plugInfo = MockConnectedPlug(c, kvmConsumerYaml, nil, "kvm")
 	s.slot, s.slotInfo = MockConnectedSlot(c, kvmCoreYaml, nil, "kvm")
+
+	// Need to Mock output of /proc/cpuinfo
+	s.tmpdir = c.MkDir()
+	dirs.SetRootDir(s.tmpdir)
```
Comment (zyga) [inline interfaces/builtin/kvm_test.go]: Please restore original root directory in `TearDownTest()`

File: interfaces/builtin/kvm.go [hunk 7]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+var procCpuinfo = "/proc/cpuinfo"
+
+func getCpuFlags() (flags string, err error) {
+	buf, err := ioutil.ReadFile(procCpuinfo)
+	if err != nil {
+		// if we can't read cpuinfo, we want to know _why_
+		return "", fmt.Errorf("error: %v", err)
+	}
+
+	pattern := []byte("\nflags\t\t:")
```
Comment (mvo5) [inline interfaces/builtin/kvm.go]: I'm slightly worried about `\t\t`, i.e. this pattern is pretty static, if /proc/flags ever gets something like `cache_alignment_size_offset_table_address : 0xdeadbeef` this may break. As much as I dislike regexp I wonder if they will be more appropriate here so that we can simply ignore the whitespace?

File: interfaces/builtin/kvm.go [hunk 8]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+	commonInterface
+}
+
+var procCpuinfo = "/proc/cpuinfo"
+
+func getCpuFlags() (flags string, err error) {
+	buf, err := ioutil.ReadFile(procCpuinfo)
+	if err != nil {
+		// if we can't read cpuinfo, we want to know _why_
+		return "", fmt.Errorf("error: %v", err)
```
Comment (mvo5) [inline interfaces/builtin/kvm.go]: Please provide a bit more context to the user here, something like: `cannot read cpu info: %v"` for the error string.

File: interfaces/builtin/kvm.go (hunk 2 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: Ack. I was duplicating procCpuinfoMinimal() in the errtracker package, but I prefer your thinking

File: interfaces/builtin/kvm.go [hunk 9]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
 
 var kvmConnectedPlugUDev = []string{`KERNEL=="kvm"`}
 
+type kvmInterface struct {
+	commonInterface
+}
+
+var procCpuinfo = "/proc/cpuinfo"
+
+func getCpuFlags() (flags string, err error) {
```
Comment (mvo5) [inline interfaces/builtin/kvm.go]: How about returning a list of strings here with the flags? Something like: `func getCpuFlags() (flags []string, err error` ? We have strutil.ListContains() (and the same for sorted lists) as a helper so that you can easily test if a string is in the list. Or even a map[string]bool  (or interface{}) ? Of course with the list approach once the flags are identified an additional `flags = strings.Fields(flagsString)` will be required but I think thats not too bad.

File: interfaces/builtin/kvm.go (hunk 3 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: I need to double-check i386 (think it uses same cpu flag and kernel module), but x86_64 and amd64 are covered.

File: interfaces/builtin/kvm.go (hunk 2 above)
Comment (mvo5) [reply interfaces/builtin/kvm.go]: Yeah, or if we go with the []string just `nil` for the empty list.

File: interfaces/builtin/kvm.go [hunk 10]
```diff
@@ -38,8 +48,60 @@ const kvmConnectedPlugAppArmor = `
...
+	return string(buf[idx+1:]), nil
+}
+
+func (iface *kvmInterface) KModConnectedPlug(spec *kmod.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Check CPU capabilities to load suitable module
+	// TODO: this only considering x86 CPUs, but some ARM, PPC and S390 CPUs also support KVM
+	m := "kvm"
+	cpu_flags, err := getCpuFlags()
+	if err != nil {
+		fmt.Println("kvm: fetching cpu info failed:", err)
```
Comment (mvo5) [inline interfaces/builtin/kvm.go]: If we want to print anything (and not error here) we should either push this via the logger or generate a snapd warning.

File: interfaces/builtin/kvm.go (hunk 7 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: I share your discomfort, but I was also avoiding regex :) I'll see what I can do

File: interfaces/builtin/kvm.go (hunk 10 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: Ack. /me confesses he didn't notice the logger implementation.

File: interfaces/builtin/kvm.go (hunk 1 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: Replaced with regexp, and function returns []string, this is gone

File: interfaces/builtin/kvm.go (hunk 3 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: vmx is available on some i386. Happy to skip the others. Comment updated to be clearer

File: interfaces/builtin/kvm.go (hunk 5 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: Done, Please see if you agree with what logger priority I use.

File: interfaces/builtin/kvm_test.go (hunk 6 above)
Comment (gerboland) [reply interfaces/builtin/kvm_test.go]: Done

File: interfaces/builtin/kvm.go (hunk 7 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: I'm using regexp, is cleaner and not in a hot code path.

File: interfaces/builtin/kvm.go (hunk 9 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: Done

File: interfaces/builtin/kvm.go (hunk 10 above)
Comment (gerboland) [reply interfaces/builtin/kvm.go]: Done, but please check if you agree with the logger method I use

File: interfaces/builtin/kvm.go [hunk 11]
```diff
@@ -38,8 +50,58 @@ const kvmConnectedPlugAppArmor = `
...
+
+func getCpuFlags() (flags []string, err error) {
+	buf, err := ioutil.ReadFile(procCpuinfo)
+	if err != nil {
+		// if we can't read cpuinfo, we want to know _why_
+		return nil, fmt.Errorf("unable to read %v: %v", procCpuinfo, err)
+	}
+
+	// want to capture the text after 'flags:' entry
+	pattern := regexp.MustCompile(`(?m)^flags\s+:\s+(.*)$`)
```
Comment (bboozzoo) [inline interfaces/builtin/kvm.go]: This should probably be package level `var` so that we don't compile the regex on each call.

Comment (gerboland) [reply interfaces/builtin/kvm.go]: Done

File: interfaces/builtin/kvm.go [hunk 12]
```diff
@@ -38,8 +50,58 @@ const kvmConnectedPlugAppArmor = `
...
+	return cpu_flags, nil
+}
+
+func (iface *kvmInterface) KModConnectedPlug(spec *kmod.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Check CPU capabilities to load suitable module
+	// NOTE: this only considers i386, x86_64 and amd64 CPUs, but some ARM, PPC and S390 CPUs also support KVM
+	m := "kvm"
+	cpu_flags, err := getCpuFlags()
+	if err != nil {
+		logger.Debugf("kvm: fetching cpu info failed:", err)
```
Comment (bboozzoo) [inline interfaces/builtin/kvm.go]:
```suggestion
		logger.Debugf("kvm: fetching cpu info failed: %v", err)
```

File: interfaces/builtin/kvm_test.go [hunk 13]
```diff
@@ -57,8 +66,25 @@ slots:
...
+	s.AddCleanup(func() { dirs.SetRootDir("/") })
+
+	mockCpuinfo := filepath.Join(s.tmpdir, "cpuinfo")
+	c.Assert(ioutil.WriteFile(mockCpuinfo, []byte(`
```
Comment (bboozzoo) [inline interfaces/builtin/kvm_test.go]: I think the PR could use another unit test where `cpuinfo` is empty or doesn't exist.

Comment (gerboland) [reply interfaces/builtin/kvm_test.go]: Done

**Comment by mvo5** [comment]
Thanks for your work on this!
